ui/src/app/workflows/components/workflow-timeline/workflow-timeline.tsx

```diff
@@ -18,128 +18,124 @@ interface WorkflowTimelineProps {
+    const [parentWidth, setParentWidth] = useState(0);
+    const [now, setNow] = useState(moment());
 
-    constructor(props: WorkflowTimelineProps) {
-        super(props);
-        this.state = {parentWidth: 0, now: moment()};
-        this.ensureRunningWorkflowRefreshing(props.workflow);
-    }
+    const containerRef = useRef(null);
 
-    public componentDidMount() {
-        this.resizeSubscription = fromEvent(window, 'resize').subscribe(() => this.updateWidth());
-        this.updateWidth();
-    }
+    useEffect(() => {
+        resizeSubscription = fromEvent(window, 'resize').subscribe(updateWidth);
+        updateWidth();
 
-    public componentWillReceiveProps(nextProps: WorkflowTimelineProps) {
-        this.ensureRunningWorkflowRefreshing(nextProps.workflow);
-    }
+        return () => {
+            if (resizeSubscription) {
+                resizeSubscription.unsubscribe();
+            }
+            if (refreshSubscription) {
+                refreshSubscription.unsubscribe();
+            }
+        };
+    }, []);
 
-    public componentWillUnmount() {
-        if (this.resizeSubscription) {
-            this.resizeSubscription.unsubscribe();
-            this.resizeSubscription = null;
-        }
-        if (this.refreshSubscription) {
-            this.refreshSubscription.unsubscribe();
-            this.refreshSubscription = null;
+    useEffect(() => {
+        ensureRunningWorkflowRefreshing(props.workflow);
+    }, [props.workflow]);
+
+    function updateWidth() {
+        if (containerRef.current) {
+            setParentWidth((containerRef.current.offsetParent || window.document.body).clientWidth - NODE_NAME_WIDTH);
         }
     }
 
-    public render() {
-        if (!this.props.workflow.status.nodes) {
-            return <p>No nodes</p>;
-        }
-        const nodes = Object.keys(this.props.workflow.status.nodes)
-            .map(id => {
-                const node = this.props.workflow.status.nodes[id];
-                node.finishedAt = node.finishedAt || this.state.now.format();
-                node.startedAt = node.startedAt || this.state.now.format();
-                return node;
-            })
-            .filter(node => node.startedAt && node.type === 'Pod')
-            .sort((first, second) => {
-                const diff = moment(first.startedAt).diff(second.startedAt);
-                // If node started almost at the same time then sort by end time
-                if (diff <= 2) {
-                    return moment(first.finishedAt).diff(second.finishedAt);
-                }
-                return diff;
+    function ensureRunningWorkflowRefreshing(workflow: models.Workflow) {
+        const completedPhases = [models.NODE_PHASE.ERROR, models.NODE_PHASE.SUCCEEDED, models.NODE_PHASE.SKIPPED, models.NODE_PHASE.OMITTED, models.NODE_PHASE.FAILED];
+        const isCompleted = workflow && workflow.status && completedPhases.indexOf(workflow.status.phase) > -1;
```

Comment:
```suggestion
        const isCompleted = workflow && workflow.status && completedPhases.includes(workflow.status.phase);
```

can use the newer `includes` function now